### ISSUE TITLE
`EventDispatcher` `set_handler` and `remove_handler` do not use `WeakMethod`

### ISSUE THREAD

**Issue by cspotcode** [pyglet/pyglet#869]
**Describe the bug**

`set_handler` and `remove_handler` do not wrap methods in `WeakMethod`.  There are a few implications from this:

- Using `set_handler` to register a handler method creates strong reference
- Using `set_handler` to add, then `remove_handlers` (plural) to remove, will fail.  Because `WeakMethod(self.method) != self.method`
- Switching between `set_handler` to set_handlers` changes this behavior.  From user's perspective, this is confusing.  The API and docs give the impression that this is a stylistic choice, not a behavioral change.

**System Information:**

**How To Reproduce**

```python
from pyglet.event import EventDispatcher

class MyEventDispatcher(EventDispatcher): pass
MyEventDispatcher.register_event_type('on_foo')

class Observer:
    def on_foo(self):
        print('Observer.on_foo called')

    def log_foo_event(self):
        print('Observer.log_foo_event called')

print('Use-case A: When handler method and event names do not match, encouraging use of set_handler')
e = MyEventDispatcher()
o = Observer()
e.push_handlers()
e.set_handler('on_foo', o.log_foo_event)
e.dispatch_event('on_foo')
o = None
# Error: o is strongly held, handler will fire again
e.dispatch_event('on_foo')

print('---')

print('Use-case B: Subscribe using a different technique than unsubscribe')
e = MyEventDispatcher()
o = Observer()
e.push_handlers(o)
e.dispatch_event('on_foo')
e.remove_handler('on_foo', o.on_foo)
# Error: handler was not removed
e.dispatch_event('on_foo')

print('---')
```

**Comment by benmoran56** [comment]
Thanks for opening the ticket. There has already been a bit of work on refactoring the `event` module, but it has not yet been merged due to some issues with references:  https://github.com/pyglet/pyglet/tree/event_refactor
This was initially stated for 2.0, but got put on the back burner.

**Comment by cspotcode** [comment]
Thanks, is there an issue or PR tracking this work?  It could serve as a description of the intended changes.  As-is, it's not clear whether or not the refactor would change the behavior described in this issue.

**Comment by benmoran56** [comment]
The original pull request is here, which includes a description of the changes. 
It actually _was_ merged, but had to be revered at the last minute before pyglet 2.0 was released (so it didn't block the release). 

The main issue (if I recall correctly), is that it was keeping some hard references that were preventing garbage collection.

**Comment by benmoran56** [comment]
@cspotcode As a quick immediate fix, the following should do it: 

```py
def my_set_handler(self, name, handler):
        if type(self._event_stack) is tuple:
            self._event_stack = [{}]

        if inspect.ismethod(handler):
            handler = WeakMethod(handler, partial(self._remove_handler, name))

        self._event_stack[0][name] = handler

EventDispatcher.set_handler = my_set_handler
```
If you have a chance, can you see if that fixes it for you as well?

**Comment by shenjackyuanjie** [comment]
a little ping, is this issue still exist?

**Comment by benmoran56** [comment]
Yes, this has not yet been fixed. It got overlooked :( but shouldn't be too difficult.
